File: src/lib/api-client.ts
import type {
  AiConfig,
  AppMeta,
  AssistantMessage,
  ChatMessage,
  ChatSession,
  LocalUser,
  PersonalSkillDictionary,
  Quest,
  QuestCompletion,
  Skill,
  UserSettings,
} from '@/domain/types'
import type {
  ActivitySession,
  ActionLogDeletionRequest,
  DailyActivityLog,
  Device,
  OpenLoop,
  PrivacyRule,
  RawEvent,
  WeeklyActivityReview,
} from '@/domain/action-log-types'
import { getIdToken } from '@/lib/auth'

const BASE_URL = import.meta.env.VITE_API_BASE_URL as string

export class ApiError extends Error {
  status: number
  path: string
  body?: unknown

  constructor(path: string, status: number, body?: unknown) {
    super(`API error: ${status} ${path}`)
    this.name = 'ApiError'
    this.status = status
    this.path = path
    this.body = body
  }
}

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const token = await getIdToken()
  const headers: Record<string, string> = { 'Content-Type': 'application/json' }
  if (token) headers['Authorization'] = `Bearer ${token}`
  const res = await fetch(`${BASE_URL}${path}`, {
    headers,
    ...options,
  })
  const raw = await res.text()
  const body = raw
    ? (() => {
        try {
          return JSON.parse(raw)
        } catch {
          return raw
        }
      })()
    : undefined
  if (!res.ok) {
    throw new ApiError(path, res.status, body)
  }
  return body as T
}

// ユーザー
export function getUser() {
  return request<LocalUser | null>('/user')
}

export function putUser(user: Partial<LocalUser>) {
  return request<{ updated: boolean }>('/user', {
    method: 'PUT',
    body: JSON.stringify(user),
  })
}

// クエスト
export function getQuests() {
  return request<Quest[]>('/quests')
}

export function postQuest(quest: Quest) {
  return request<Quest>('/quests', {
    method: 'POST',
    body: JSON.stringify(quest),
  })
}

export function putQuest(id: string, quest: Partial<Quest>) {
  return request<Quest>(`/quests/${id}`, {
    method: 'PUT',
    body: JSON.stringify(quest),
  })
}

export function deleteQuest(id: string) {
  return request<{ deleted: string }>(`/quests/${id}`, {
    method: 'DELETE',
  })
}

// 完了記録
export function getCompletions() {
  return request<QuestCompletion[]>('/completions')
}

export function postCompletion(completion: QuestCompletion) {
  return request<unknown>('/completions', {
    method: 'POST',
    body: JSON.stringify(completion),
  })
}

export function putCompletion(id: string, updates: Partial<QuestCompletion>) {
  return request<unknown>(`/completions/${id}`, {
    method: 'PUT',
    body: JSON.stringify(updates),
  })
}

// スキル
export function getSkills() {
  return request<Skill[]>('/skills')
}

export function postSkill(skill: Skill) {
  return request<Skill>('/skills', {
    method: 'POST',
    body: JSON.stringify(skill),
  })
}

export function putSkill(id: string, skill: Partial<Skill>) {
  return request<Skill>(`/skills/${id}`, {
    method: 'PUT',
    body: JSON.stringify(skill),
  })
}

// 設定
export function getSettings() {
  return request<UserSettings | null>('/settings')
}

export function putSettings(settings: Partial<UserSettings>) {
  return request<{ updated: boolean }>('/settings', {
    method: 'PUT',
    body: JSON.stringify(settings),
  })
}

// AI設定
export function getAiConfig() {
  return request<AiConfig | null>('/ai-config')
}

export function putAiConfig(config: Partial<AiConfig>) {
  return request<{ updated: boolean }>('/ai-config', {
    method: 'PUT',
    body: JSON.stringify(config),
  })
}

// メタ
export function getMeta() {
  return request<AppMeta | null>('/meta')
}

export function putMeta(meta: AppMeta) {
  return request<{ updated: boolean }>('/meta', {
    method: 'PUT',
    body: JSON.stringify(meta),
  })
}

// メッセージ
export function getMessages() {
  return request<AssistantMessage[]>('/messages')
}

export function postMessage(message: AssistantMessage) {
  return request<AssistantMessage>('/messages', {
    method: 'POST',
    body: JSON.stringify(message),
  })
}

// ブラウジング時間
export interface BrowsingTimeData {
  date: string
  domains: Record<string, { totalSeconds: number; category: string; isGrowth: boolean }>
  totalSeconds: number
}

export function getBrowsingTimes(from: string, to: string) {
  return request<BrowsingTimeData[]>(`/browsing-times?from=${from}&to=${to}`)
}

export interface HealthDataEntry {
  date: string
  time: string
  weight_kg: number | null
  body_fat_pct: number | null
  source?: string
}

export function getHealthData(from: string, to: string) {
  return request<HealthDataEntry[]>(`/health-data?from=${from}&to=${to}`)
}

export function postHealthData(entries: HealthDataEntry[]) {
  return request<{ synced: number }>('/health-data', {
    method: 'POST',
    body: JSON.stringify({ entries }),
  })
}

// アクティビティログ
export function postActivityLogs(entries: Array<{
  timestamp: string
  source: string
  action: string
  category: string
  details: Record<string, unknown>
}>) {
  return request<{ logged: number }>('/activity-logs', {
    method: 'POST',
    body: JSON.stringify({ entries }),
  })
}

// 辞書
export function getDictionary() {
  return request<PersonalSkillDictionary[]>('/dictionary')
}

export function postDictEntry(entry: PersonalSkillDictionary) {
  return request<PersonalSkillDictionary>('/dictionary', {
    method: 'POST',
    body: JSON.stringify(entry),
  })
}

export function putDictEntry(id: string, entry: Partial<PersonalSkillDictionary>) {
  return request<PersonalSkillDictionary>(`/dictionary/${id}`, {
    method: 'PUT',
    body: JSON.stringify(entry),
  })
}

// チャットセッション
export function getChatSessions() {
  return request<ChatSession[]>('/chat-sessions')
}

export function postChatSession(session: ChatSession) {
  return request<ChatSession>('/chat-sessions', {
    method: 'POST',
    body: JSON.stringify(session),
  })
}

export function putChatSession(id: string, updates: Partial<ChatSession>) {
  return request<ChatSession>(`/chat-sessions/${id}`, {
    method: 'PUT',
    body: JSON.stringify(updates),
  })
}

export function deleteChatSession(id: string) {
  return request<{ deleted: string }>(`/chat-sessions/${id}`, {
    method: 'DELETE',
  })
}

// チャットメッセージ
export function getChatMessages(sessionId: string) {
  return request<ChatMessage[]>(`/chat-sessions/${sessionId}/messages`)
}

export function getChatMessagesRange(from: string, to: string) {
  return request<ChatMessage[]>(`/chat-messages?from=${from}&to=${to}`)
}

export function postChatMessage(sessionId: string, message: ChatMessage) {
  return request<ChatMessage>(`/chat-sessions/${sessionId}/messages`, {
    method: 'POST',
    body: JSON.stringify(message),
  })
}

// アクティビティログ取得
export interface ActivityLogEntry {
  timestamp: string
  source: string
  action: string
  category: string
  details: Record<string, unknown>
}

export function getActivityLogs(from: string, to: string) {
  return request<ActivityLogEntry[]>(`/activity-logs?from=${from}&to=${to}`)
}

export function postActionLogRawEvents(input: { deviceId: string; events: RawEvent[] }) {
  return request<{ logged: number }>('/action-log/raw-events', {
    method: 'POST',
    body: JSON.stringify(input),
  })
}

export function getActionLogRawEvents(from: string, to: string) {
  return request<RawEvent[]>(`/action-log/raw-events?from=${from}&to=${to}`)
}

export function putActionLogSessions(input: { deviceId: string; sessions: ActivitySession[] }) {
  return request<{ updated: number }>('/action-log/sessions', {
    method: 'PUT',
    body: JSON.stringify(input),
  })
}

export function putActionLogSessionHidden(
  id: string,
  input: { dateKey: string; hidden: boolean },
) {
  return request<ActivitySession>(`/action-log/sessions/${id}/hidden`, {
    method: 'PUT',
    body: JSON.stringify(input),
  })
}

export function getActionLogSessions(from: string, to: string) {
  return request<ActivitySession[]>(`/action-log/sessions?from=${from}&to=${to}`)
}

export function getActionLogDailyActivityLogs(from: string, to: string) {
  return request<DailyActivityLog[]>(`/action-log/daily?from=${from}&to=${to}`)
}

export function getActionLogDailyActivityLog(dateKey: string) {
  return request<DailyActivityLog | null>(`/action-log/daily/${dateKey}`)
}

export function putActionLogDailyActivityLog(log: DailyActivityLog) {
  return request<DailyActivityLog>(`/action-log/daily/${log.dateKey}`, {
    method: 'PUT',
    body: JSON.stringify(log),
  })
}

export function getActionLogWeeklyActivityReview(weekKey: string) {
  return request<WeeklyActivityReview | null>(`/action-log/weekly/${weekKey}`)
}

export function getActionLogWeeklyActivityReviews(year: number) {
  return request<WeeklyActivityReview[]>(`/action-log/weekly?year=${year}`)
}

export function putActionLogWeeklyActivityReview(review: WeeklyActivityReview) {
  return request<WeeklyActivityReview>(`/action-log/weekly/${review.weekKey}`, {
    method: 'PUT',
    body: JSON.stringify(review),
  })
}

export function getActionLogDevices() {
  return request<Device[]>('/action-log/devices')
}

export function putActionLogDevice(id: string, updates: Partial<Device>) {
  return request<Device>(`/action-log/devices/${id}`, {
    method: 'PUT',
    body: JSON.stringify(updates),
  })
}

export function getActionLogPrivacyRules() {
  return request<PrivacyRule[]>('/action-log/privacy-rules')
}

export function putActionLogPrivacyRules(rules: PrivacyRule[]) {
  return request<{ updated: number }>('/action-log/privacy-rules', {
    method: 'PUT',
    body: JSON.stringify({ rules }),
  })
}

export function getActionLogOpenLoops(from: string, to: string) {
  return request<OpenLoop[]>(`/action-log/open-loops?from=${from}&to=${to}`)
}

export function putActionLogOpenLoops(input: { dateKeys: string[]; openLoops: OpenLoop[] }) {
  return request<{ updated: number }>('/action-log/open-loops', {
    method: 'PUT',
    body: JSON.stringify(input),
  })
}

export function deleteActionLogRange(from: string, to: string) {
  return request<{
    deleted: {
      rawEvents: number
      sessions: number
      dailyLogs: number
      weeklyReviews: number
      openLoops: number
    }
    deletionRequestId: string
  }>(`/action-log/range?from=${from}&to=${to}`, {
    method: 'DELETE',
  })
}

export function getActionLogDeletionRequests() {
  return request<ActionLogDeletionRequest[]>('/action-log/deletion-requests')
}

export function postActionLogDeletionRequestAck(id: string) {
  return request<{ acked: string }>(`/action-log/deletion-requests/${id}/ack`, {
    method: 'POST',
  })
}

// ---- 状況ログ ----

export type SituationLogEntry = {
  summary: string
  timestamp: string
  details: {
    camera_summaries?: string[]
    desktop_summaries?: string[]
    active_apps?: string[]
  }
}

export function getSituationLogs(from: string, to: string) {
  return request<SituationLogEntry[]>(`/situation-logs?from=${from}&to=${to}`)
}

// ---- 栄養素 ----

import type { MealType, NutritionRecord } from '@/domain/types'

export type NutritionDayResult = Record<MealType, NutritionRecord | null>

export function getNutrition(date: string) {
  return request<NutritionDayResult>(`/nutrition?date=${date}`)
}

export type NutritionRangeResult = Record<string, NutritionDayResult>

export function getNutritionRange(from: string, to: string) {
  return request<NutritionRangeResult>(`/nutrition?from=${from}&to=${to}`)
}

export function putNutrition(date: string, mealType: MealType, record: Omit<NutritionRecord, 'userId'>) {
  return request<NutritionRecord>(`/nutrition/${date}/${mealType}`, {
    method: 'PUT',
    body: JSON.stringify(record),
  })
}

// ---- Fitbit ----

export interface FitbitHeartZone {
  name: string
  min: number
  max: number
  minutes: number
  calories_out: number
}

export interface FitbitSummary {
  date: string
  heart: {
    resting_heart_rate: number | null
    intraday_points: number
    heart_zones: FitbitHeartZone[]
  } | null
  active_zone_minutes: {
    intraday_points: number
    minutes_total_estimate: number | null
    summary_rows: number
  } | null
  sleep: {
    main_sleep: {
      date_of_sleep: string
      start_time: string
      end_time: string
      minutes_asleep: number
      minutes_awake: number
      time_in_bed: number
      deep_minutes: number | null
      light_minutes: number | null
      rem_minutes: number | null
      wake_minutes: number | null
    } | null
    all_sleep_count: number
  } | null
  activity: {
    steps: number | null
    distance: number | null
    calories: number | null
    very_active_minutes: number | null
    fairly_active_minutes: number | null
    lightly_active_minutes: number | null
    sedentary_minutes: number | null
  } | null
}

export function getFitbitData(from: string, to: string) {
  return request<FitbitSummary[]>(`/fitbit-data?from=${from}&to=${to}`)
}

File: lily_desktop/api/api_client.py
from __future__ import annotations

import logging
from typing import Any

import httpx

from api.auth import CognitoAuth
from core.constants import API_BASE_URL

logger = logging.getLogger(__name__)


class ApiClient:
    """既存API Gatewayへのクライアント (src/lib/api-client.ts の Python移植)"""

    def __init__(self, auth: CognitoAuth):
        self._auth = auth
        self._http = httpx.AsyncClient(timeout=30.0)
        self._base = API_BASE_URL

    async def _request(
        self, method: str, path: str, json: Any = None, params: dict | None = None
    ) -> Any:
        token = await self._auth.get_id_token()
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        }
        resp = await self._http.request(
            method, f"{self._base}{path}", headers=headers, json=json, params=params
        )
        resp.raise_for_status()
        return resp.json()

    # ---- ユーザー ----
    async def get_user(self) -> dict | None:
        return await self._request("GET", "/user")

    # ---- クエスト ----
    async def get_quests(self) -> list[dict]:
        return await self._request("GET", "/quests")

    async def post_quest(self, quest: dict) -> dict:
        return await self._request("POST", "/quests", json=quest)

    async def delete_quest(self, quest_id: str) -> dict:
        return await self._request("DELETE", f"/quests/{quest_id}")

    async def put_quest(self, quest_id: str, updates: dict) -> dict:
        return await self._request("PUT", f"/quests/{quest_id}", json=updates)

    # ---- 完了記録 ----
    async def get_completions(self) -> list[dict]:
        return await self._request("GET", "/completions")

    async def post_completion(self, completion: dict) -> dict:
        return await self._request("POST", "/completions", json=completion)

    async def put_completion(self, completion_id: str, updates: dict) -> dict:
        return await self._request("PUT", f"/completions/{completion_id}", json=updates)

    # ---- スキル ----
    async def get_skills(self) -> list[dict]:
        return await self._request("GET", "/skills")

    async def post_skill(self, skill: dict) -> dict:
        return await self._request("POST", "/skills", json=skill)

    # ---- 設定 ----
    async def get_settings(self) -> dict | None:
        return await self._request("GET", "/settings")

    # ---- AI設定 ----
    async def get_ai_config(self) -> dict | None:
        return await self._request("GET", "/ai-config")

    # ---- メタ ----
    async def get_meta(self) -> dict | None:
        return await self._request("GET", "/meta")

    # ---- メッセージ ----
    async def get_messages(self) -> list[dict]:
        return await self._request("GET", "/messages")

    # ---- ブラウジング時間 ----
    async def get_browsing_times(self, from_date: str, to_date: str) -> list[dict]:
        return await self._request(
            "GET", "/browsing-times", params={"from": from_date, "to": to_date}
        )

    # ---- 体重・体脂肪率 ----
    async def get_health_data(self, from_date: str, to_date: str) -> list[dict]:
        return await self._request(
            "GET", "/health-data", params={"from": from_date, "to": to_date}
        )

    async def post_health_data(self, entries: list[dict]) -> dict:
        return await self._request("POST", "/health-data", json={"entries": entries})

    # ---- Fitbit ----
    async def post_fitbit_data(self, summary: dict) -> dict:
        return await self._request("POST", "/fitbit-data", json=summary)

    async def get_fitbit_data(self, from_date: str, to_date: str) -> list[dict]:
        return await self._request(
            "GET", "/fitbit-data", params={"from": from_date, "to": to_date}
        )

    # ---- アクティビティログ ----
    async def get_activity_logs(self, from_date: str, to_date: str) -> list[dict]:
        return await self._request(
            "GET", "/activity-logs", params={"from": from_date, "to": to_date}
        )

    async def post_action_log_raw_events(self, payload: dict) -> dict:
        return await self._request("POST", "/action-log/raw-events", json=payload)

    async def get_action_log_raw_events(
        self, from_date: str, to_date: str
    ) -> list[dict]:
        return await self._request(
            "GET", "/action-log/raw-events", params={"from": from_date, "to": to_date}
        )

    async def get_action_log_sessions(
        self, from_date: str, to_date: str
    ) -> list[dict]:
        return await self._request(
            "GET", "/action-log/sessions", params={"from": from_date, "to": to_date}
        )

    async def put_action_log_sessions(self, payload: dict) -> dict:
        return await self._request("PUT", "/action-log/sessions", json=payload)

    async def put_action_log_session_hidden(
        self, session_id: str, payload: dict
    ) -> dict:
        return await self._request(
            "PUT", f"/action-log/sessions/{session_id}/hidden", json=payload
        )

    async def get_action_log_daily_logs(
        self, from_date: str, to_date: str
    ) -> list[dict]:
        return await self._request(
            "GET", "/action-log/daily", params={"from": from_date, "to": to_date}
        )

    async def get_action_log_daily_log(self, date_key: str) -> dict | None:
        return await self._request("GET", f"/action-log/daily/{date_key}")

    async def put_action_log_daily_log(self, log: dict) -> dict:
        return await self._request(
            "PUT", f"/action-log/daily/{log['dateKey']}", json=log
        )

    async def get_action_log_weekly_reviews(self, year: int) -> list[dict]:
        return await self._request(
            "GET", "/action-log/weekly", params={"year": str(year)}
        )

    async def get_action_log_weekly_review(self, week_key: str) -> dict | None:
        return await self._request("GET", f"/action-log/weekly/{week_key}")

    async def put_action_log_weekly_review(self, review: dict) -> dict:
        return await self._request(
            "PUT", f"/action-log/weekly/{review['weekKey']}", json=review
        )

    async def get_action_log_devices(self) -> list[dict]:
        return await self._request("GET", "/action-log/devices")

    async def put_action_log_device(self, device_id: str, updates: dict) -> dict:
        return await self._request(
            "PUT", f"/action-log/devices/{device_id}", json=updates
        )

    async def get_action_log_privacy_rules(self) -> list[dict]:
        return await self._request("GET", "/action-log/privacy-rules")

    async def put_action_log_privacy_rules(self, rules: list[dict]) -> dict:
        return await self._request(
            "PUT", "/action-log/privacy-rules", json={"rules": rules}
        )

    async def get_action_log_open_loops(
        self, from_date: str, to_date: str
    ) -> list[dict]:
        return await self._request(
            "GET", "/action-log/open-loops", params={"from": from_date, "to": to_date}
        )

    async def put_action_log_open_loops(self, payload: dict) -> dict:
        return await self._request("PUT", "/action-log/open-loops", json=payload)

    async def delete_action_log_range(self, from_date: str, to_date: str) -> dict:
        return await self._request(
            "DELETE", "/action-log/range", params={"from": from_date, "to": to_date}
        )

    async def get_action_log_deletion_requests(self) -> list[dict]:
        return await self._request("GET", "/action-log/deletion-requests")

    async def ack_action_log_deletion_request(self, request_id: str) -> dict:
        return await self._request(
            "POST", f"/action-log/deletion-requests/{request_id}/ack"
        )

    # ---- 栄養素 ----
    async def get_nutrition_range(self, from_date: str, to_date: str) -> dict:
        return await self._request(
            "GET", "/nutrition", params={"from": from_date, "to": to_date}
        )

    # ---- 状況ログ ----
    async def get_situation_logs(self, from_date: str, to_date: str) -> list[dict]:
        return await self._request(
            "GET", "/situation-logs", params={"from": from_date, "to": to_date}
        )

    async def post_situation_log(self, log: dict) -> dict:
        return await self._request("POST", "/situation-logs", json=log)

    # ---- 辞書 ----
    async def get_dictionary(self) -> list[dict]:
        return await self._request("GET", "/dictionary")

    # ---- チャットセッション ----
    async def get_chat_sessions(self) -> list[dict]:
        return await self._request("GET", "/chat-sessions")

    async def post_chat_session(self, session: dict) -> dict:
        return await self._request("POST", "/chat-sessions", json=session)

    async def put_chat_session(self, session_id: str, updates: dict) -> dict:
        return await self._request(
            "PUT", f"/chat-sessions/{session_id}", json=updates
        )

    async def delete_chat_session(self, session_id: str) -> dict:
        return await self._request("DELETE", f"/chat-sessions/{session_id}")

    # ---- チャットメッセージ ----
    async def get_chat_messages(self, session_id: str) -> list[dict]:
        return await self._request(
            "GET", f"/chat-sessions/{session_id}/messages"
        )

    async def get_chat_messages_range(self, from_date: str, to_date: str) -> list[dict]:
        return await self._request(
            "GET",
            "/chat-messages",
            params={"from": from_date, "to": to_date},
        )

    async def post_chat_message(self, session_id: str, message: dict) -> dict:
        return await self._request(
            "POST", f"/chat-sessions/{session_id}/messages", json=message
        )

    async def close(self) -> None:
        await self._http.aclose()
